docs/source/jobs.rst

```diff
@@ -239,41 +239,137 @@ Use cases for the **job document** include, but are not limited to:
+For cases where `job`-associated data may be accessed from multiple sources at the same time or other instances where multiple files may be preferred to one large file, :py:attr:`Job.stores` should be used instead of :py:attr:`Job.data`.
+Further discussion of :py:attr:`Job.stores` is provided in the folling section, Job Stores.
+
+Reading and Writing data
+-----------------------
+
 An example of storing data:
 
 .. code-block:: python
 
     >>> import numpy as np
     >>> job = project.open_job(statepoint)
-    >>> job.data['x'] = np.arange(0, 1, 0.01)
-    >>> job.data['hello'] = 'world'
+    >>> job.data['x'] = np.ones([10, 3, 4])
+
 
 Just like the job *state point* and *document*, individual keys may be accessed either as attributes or through a functional interface, *e.g.*.
-The following examples are all equivalent:
+
+To access data as attributes:
 
 .. code-block:: python
 
-    >>> print(job.data.get('hello'))
-    world
-    >>> print(job.data['hello'])
-    world
-    >>> print(job.data.hello)
-    world
+    >>> with job.data:
+    ...     x = job.data.x[:]
+
+To access data as a key:
+
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data['x'][:]
+
+Through a functional interface:
+
+.. code-block:: python
+    
+    >>> with job.data:
+    ...     x = job.data.get('x')[:]
 
 .. tip::
 
-     Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's `built-in dictionaries <https://docs.python.org/3/library/stdtypes.html#dict.get>`_.
+     Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's `built-in dictionaries      Use the :py:meth:`Job.data.get` method to return ``None`` or another specified default value for missing values. This works exactly like with python's built-in dictionaries (see :py:meth:`dict.get`). 
+    
+Accessing arrays
+----------------
+
+All values stored in :attr:`job.data` are returned as copies, except for arrays, which are accessed *by reference* and not automatically copied into memory.
+That is important to enable the storage of massive arrays that do not necessarily fit into memory.
+
+For fast and efficient data access, NumPy slicing syntax may be used to access data. 
+Here are a few examples with outputs omitted:
+   
+.. code-block:: python
+    
+    >>> with job.data:
+    ...     job.data.x[0, 0, 0]
+    ...     job.data.x[1:3, 0, :]
+    ...     job.data.x[:, 1, 3]
+    
+To load entire arrays to memory, NumPy slicing syntax may be used:
+ 
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data.x[:]
+
+You can also create an explicit memory copy of the entire array using the copy-operator ``[()]``:
+
+.. code-block:: python
+
+    >>> with job.data:
+    ...     x = job.data.x[()]
```

Comment:
Is there any difference between `[:]` and `[()]`? If not, we should probably merge these two examples, if yes, then we should point those out.

Reply:
I can't find any reference to the `[()]` syntax on the NumPy or h5py docs. From my tests, it seems to act the same as `[:]`, which is my preference. I don't think I've seen anyone but @csadorf write `[()]` but I think that `[:]` is familiar to most NumPy users.

Reply:
I don't know how these behave for h5py datasets, I've never noticed any difference but I haven't tested extensively. I believe that are some very subtle differences when using these for NumPy arrays, specifically with respect to how they behave for 0-d arrays in older versions of NumPy and whether or not the returned object is a view on the original array or if the original array itself is returned. However, since they both seem to give copies in the h5py case I agree with converging on just `[:]`, which is what is most familiar.

Reply:
I removed the `[()]` example and kept all of the `[:]` examples. I can add it back if you guys would like

Reply:
I think that's fine.

Reply:
Ok, so I did my own research. The operators are described here: http://docs.h5py.org/en/latest/high/dataset.html#reading-writing-data

According to those docs, this is the purpose of the `[()]` operator:
> An empty tuple (`()`) to retrieve all data or scalar data

This means that `[()]` and `[:]` are not identical, only the former can be used to copy **both** scalars and arrays and in that sense `[()]` is preferable as a copy-operator. I confirmed this in a minimal test locally.

@klywang Can you clarify this in the docs?

Reply:
I added some details about `[()]`. I tried to load strings with `[()]`, but I got an error so I included what each operator can and cannot be used to load.

I also inclded a tip about how `job.doc` should be used for scalars rather than `job.data`. Please let me know if that is not an intended use. I can take it off and add some more examples on accessing scalars and strings.

Reply:
@klywang Good work. Thank you for your patience and persistence in re-writing this 2 or 3 times. I think we all learned things about memory management that we didn't know when we started this documentation effort! I am happy with the outcomes and will resolve this conversation.